rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+/// Returns an error if `points` is less than 2 or if the interval is unbounded.
+pub fn spaced_points(
+    self,
+    points: usize,
+) -> Result<impl Iterator<Item = f32>, SpacedPointsError> {
+```
+
+The `Interval` type also implements `TryFrom<RangeInclusive>`, which may be desirable if you want to use
+the `start..=end` syntax. One of the primary benefits of `Interval` (in addition to these methods) is
+that it is `Copy`, so it is easy to take intervals and throw them around. 
+
+### Sampling
+
+The `Curve::sample` method is not intrinsically constrained by the curve's `domain` interval. Instead,
+implementors of `Curve<T>` are free to determine how samples drawn from outside the `domain` will behave.
+However, variants of `sample` (as well as other important methods) use the `domain` explicitly:
+```rust
+/// Sample a point on this curve at the parameter value `t`, returning `None` if the point is
+/// outside of the curve's domain.
+fn sample_checked(&self, t: f32) -> Option<T> { //... }
+
+/// Sample a point on this curve at the parameter value `t`, clamping `t` to lie inside the
+/// domain of the curve.
+fn sample_clamped(&self, t: f32) -> T { //... }
+```
+
+### The Main Curve API
+
+Now, let us turn our attention back to `Curve<T>`, which exposes a functional API similar to that of `Iterator`.
+We will explore its main components one-by-one. The first of those is `map`:
+```rust
+/// Create a new curve by mapping the values of this curve via a function `f`; i.e., if the
+/// sample at time `t` for this curve is `x`, the value at time `t` on the new curve will be
+/// `f(x)`.
+fn map<S>(self, f: impl Fn(T) -> S) -> impl Curve<S>
+where
+    Self: Sized,
+{ //... }
+```
+As you can see, `map` takes our curve and, consuming it, produces a new curve whose sample values are the sample
+values of the starting curve mapped through `f`. For example, if we started with a curve in three-dimensional space
+and wanted to project it onto the XY-plane, we could do that with `map`:
+```rust
+// A 3d curve, implementing `Curve<Vec3>`
+let my_3d_curve = function_curve(interval(0.0, 2.0).unwrap(), |t| Vec3::new(t * t, 2.0 * t, t - 1.0));
+// Its 2d projection, implementing `Curve<Vec2>`
+let my_2d_curve = my_3d_curve.map(|v| Vec2::new(vec.x, vec.y));
+```
+As you might expect, `map` is lazy like its `Iterator` counterpart, so the function it takes as input is only
+evaluated when the resulting curve is actually sampled.
+
+---
+
+Next up is `reparametrize`, another of the most important API methods:
+```rust
+/// Create a new [`Curve`] whose parameter space is related to the parameter space of this curve
+/// by `f`. For each time `t`, the sample from the new curve at time `t` is the sample from
+/// this curve at time `f(t)`. The given `domain` will be the domain of the new curve. The
+/// function `f` is expected to take `domain` into `self.domain()`.
+fn reparametrize(self, domain: Interval, f: impl Fn(f32) -> f32) -> impl Curve<T>
```

Comment:
> This function should take `Curve<f32>` as argument, such that the type system can determine continuity information for the resulting curve.

There is a version that does that (`reparametrize_by_curve`), but I kind of like the barebones nature of `reparametrize`, especially since it makes altering curve domain very easy.

> With this change, it might make sense to swap the arguments and rename the function to just `chain` (or perhaps `pipe`, but that's for the functional programmers). Chain is easier to understand, write and pronounce.

Sounds like `compose` is probably changing to `chain` — I'm not necessarily against changing the name, but `reparametrize` is also what it literally does. It would be nice if we could convey succinctly that the set of values taken by the curve is not generally changing.

> Caveat: There is no check on whether the first curve will only produce values that fall inside the domain of the second curve. For this purpose, it might be useful for Curve's to expose an additional `range(&self) -> Interval` function, though I'm not sure whether that is possible in rust.

This only really makes sense for `Curve<f32>`, but maybe it's something that could end up appearing in an extension trait later on if we find the need for it; in general, of course, this isn't something that you can compute easily.